File: tests/adapters/c/goldens/comments/complex_policy.c
/**
 * C module for testing comment optimization.
 *
 * This module contains various types of comments to test
 * different comment processing polic…
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// … comment omitted
#define MODULE_VERSION "1.0.0"  // TODO: Move to config file

/**
 * Structure with documentation comments.
 * This should be preserved when keeping documentation comments.
 */
typedef struct {
    int id;           // … comment omitted
    char* name;       // FIXME: Should validate name format
    char* email;      // … comment omitted
    // … comment omitted
    void* profile;
} User;

typedef struct {
    char* bio;
    char* avatar;
} Profile;

typedef struct {
    void* config;     // … comment omitted
    void* logger;     // … comment omitted
} CommentedService;

/**
 * Service constructor with detailed documentation.
 *
 * Initializes the service with the provided configuration
 * and sets up the loggin…
 */
CommentedService* commented_service_new(void* config, void* logger) {
    CommentedService* service = (CommentedService*)malloc(sizeof(CommentedService));
    if (!service) return NULL;

    service->config = config;
    service->logger = logger;

    // … comment omitted

    // TODO: Add configuration validation
    // FIXME: Logger should be required, not optional

    return service;
}

/**
 * Process user data with validation.
 *
 * This function performs comprehensive user data processing including
 * validation, transformation…
 */
User* process_user(User* userData) {
    // … comment omitted
    if (!userData) {
        fprintf(stderr, "User data is required\n");
        return NULL;
    }

    // … comment omitted
    int is_valid = 1;
    if (!userData->name || strlen(userData->name) == 0) {
        // … comment omitted
        fprintf(stderr, "Validation failed: name is required\n");
        is_valid = 0;
    }

    if (!userData->email || strlen(userData->email) == 0) {
        fprintf(stderr, "Validation failed: email is required\n");
        is_valid = 0;
    }

    if (!is_valid) {
        return NULL;
    }

    // … comment omitted
    User* transformed = transform_user_data(userData);

    // … comment omitted
    // NOTE: This could be optimized with batch operations
    User* saved = save_user(transformed);

    return saved;  // … comment omitted
}

static User* transform_user_data(User* userData) {
    // … comment omitted
    User* user = (User*)malloc(sizeof(User));
    if (!user) return NULL;

    user->id = generate_user_id();      // … comment omitted
    user->name = strdup(userData->name); // … comment omitted
    user->email = strdup(userData->email); // … comment omitted
    user->profile = userData->profile ? userData->profile : NULL;

    return user;
}

/**
 * Generate unique user ID.
 * @return Generated user ID
 */
static int generate_user_id(void) {
    // … comment omitted
    return rand() % 1000000;
}

// TODO: Implement proper persistence layer
static User* save_user(User* user) {
    // … comment omitted
    if (user) {
        printf("Saving user: %d\n", user->id);
    }

    // … comment omitted

    return user;  // … comment omitted
}

/**
 * Utility function with comprehensive documentation.
 *
 * @param input The input string to process
 * @return Processed string res…
 */
char* process_string(const char* input) {
    // … comment omitted
    if (!input || strlen(input) == 0) {
        return strdup("");  // … comment omitted
    }

    // … comment omitted
    size_t len = strlen(input);
    char* result = (char*)malloc(len + 1);
    if (!result) return NULL;

    // … comment omitted
    strcpy(result, input);

    return result;  // … comment omitted
}

// … comment omitted
static void undocumented_helper(void) {
    // … comment omitted
    const char* data = "helper data";

    // … comment omitted
    printf("%s\n", data);  // … comment omitted
}

// … comment omitted
typedef struct {
    int is_valid;      // … comment omitted
    char** errors;     // … comment omitted
    int error_count;   // … comment omitted
} ValidationResult;

typedef struct {
    int timeout;       // … comment omitted
    int retries;       // … comment omitted
    char* base_url;    // … comment omitted
} ServiceConfig;

// … comment omitted
ServiceConfig DEFAULT_CONFIG = {
    5000,              // … comment omitted
    3,                 // … comment omitted
    "http://localhost:3000"  // … comment omitted
};

File: tests/adapters/rust/goldens/comments/complex_policy.rs
//! Rust module for testing comment optimization.
//!
//! This module contains various types of comments to test
//! different comment processing policies and edge cases.

use std::fmt;
use regex::Regex;

// … comment omitted
const MODULE_VERSION: &str = "1.0.0"; // TODO: Move to config file

/// User represents a user with documentation comments.
/// This should be preserved when keeping documentation comments.
#[derive(Debug, Clone)]
struct User {
    id: i32,            // … comment omitted
    name: String,       // FIXME: Should validate name format
    email: String,      // … comment omitted
    profile: Option<Profile>, // … comment omitted
}

/// Profile contains user profile information
#[derive(Debug, Clone)]
struct Profile {
    bio: String,
    avatar: String,
}

/// CommentedService provides various operations with extensive comments
struct CommentedService {
    config: Box<dyn std::any::Any>, // … comment omitted
    logger: Option<Box<dyn std::any::Any>>, // … comment omitted
}

impl CommentedService {
    /// Creates a new service instance with detailed documentation.
    ///
    /// Initializes the service with the provided configuration
    /// and sets up the logging system if logger is provided.
    fn new(config: Box<dyn std::any::Any>, logger: Option<Box<dyn std::any::Any>>) -> Self {
        let service = Self { config, logger };

        // … comment omitted

        // TODO: Add configuration validation
        // FIXME: Logger should be required, not optional

        service
    }

    /// Processes user data with validation.
    ///
    /// This function performs comprehensive user data processing including
    /// validation, transformation, and persistence operations. It handles
    /// various edge cases and provides detailed error reporting.
    fn process_user(&self, user_data: User) -> Result<User, String> {
        // … comment omitted
        if user_data.name.is_empty() {
            return Err("User data is required".to_string());
        }

        // … comment omitted (5 lines)
        let validation_result = self.validate_user(&user_data);
        if !validation_result.is_valid {
            // … comment omitted
            eprintln!("Validation failed: {:?}", validation_result.errors);
            return Err("Validation failed".to_string());
        }

        // … comment omitted
        let transformed = self.transform_user_data(user_data);

        // … comment omitted
        // NOTE: This could be optimized with batch operations
        let saved = self.save_user(transformed)?;

        Ok(saved) // … comment omitted
    }

    fn validate_user(&self, user_data: &User) -> ValidationResult {
        // … comment omitted
        let mut errors = Vec::new();

        // … comment omitted
        if user_data.name.is_empty() {
            errors.push("Name is required".to_string()); // … comment omitted
        }

        if user_data.email.is_empty() {
            errors.push("Email is required".to_string());
        }

        // … comment omitted
        let email_regex = Regex::new(r"^[^\s@]+@[^\s@]+\.[^\s@]+$").unwrap();
        if !user_data.email.is_empty() && !email_regex.is_match(&user_data.email) {
            errors.push("Invalid email format".to_string());
        }

        ValidationResult {
            is_valid: errors.is_empty(),
            errors,
        }
    }

    fn transform_user_data(&self, user_data: User) -> User {
        // … comment omitted (5 lines)
        User {
            id: self.generate_user_id(),           // … comment omitted
            name: user_data.name.trim().to_string(),  // … comment omitted
            email: user_data.email.to_lowercase(), // … comment omitted
            profile: user_data.profile,            // … comment omitted
        }
    }

    /// Generates unique user ID.
    fn generate_user_id(&self) -> i32 {
        // … comment omitted
        use std::time::{SystemTime, UNIX_EPOCH};
        let duration = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
        (duration.as_secs() % 1_000_000) as i32
    }

    // TODO: Implement proper persistence layer
    fn save_user(&self, user: User) -> Result<User, String> {
        // … comment omitted
        println!("Saving user: {}", user.id);

        // … comment omitted

        Ok(user) // … comment omitted
    }
}

/// Utility function with comprehensive documentation.
///
/// It processes the input string according to specific rules.
fn process_string(input: &str) -> String {
    // … comment omitted
    if input.is_empty() {
        return String::new(); // … comment omitted
    }

    // … comment omitted (5 lines)
    let trimmed = input.trim();
    let lowercase = trimmed.to_lowercase();
    let cleaned = lowercase.chars().filter(|c| c.is_alphanumeric() || c.is_whitespace()).collect();

    cleaned // … comment omitted
}

// … comment omitted
fn undocumented_helper() {
    // … comment omitted
    let data = "helper data";

    // … comment omitted
    println!("{}", data); // … comment omitted
}

/// ValidationResult holds validation results
struct ValidationResult {
    is_valid: bool,     // … comment omitted
    errors: Vec<String>, // … comment omitted
}

/// ServiceConfig holds service configuration
struct ServiceConfig {
    timeout: i32,    // … comment omitted
    retries: i32,    // … comment omitted
    base_url: String, // … comment omitted
}

// … comment omitted (4 lines)
const DEFAULT_CONFIG: ServiceConfig = ServiceConfig {
    timeout: 5000,                    // … comment omitted
    retries: 3,                       // … comment omitted
    base_url: String::from("http://localhost:3000"), // … comment omitted
};
